### ISSUE TITLE
[basic.compound] Current wording seems to disallow formation of pointer to static member function and explicit object member function without `&` operator and 

### ISSUE THREAD

**Issue by ranaanoop** [cplusplus/CWG#572]
Full name of submitter: Anoop Rana

Reference (section label): [basic.compound]

Issue description:

I've noticed several issue regarding pointer to member. I'll start by describing one of them here in this issue. 

In particular, [basic.compound](https://eel.is/c++draft/basic.compound#1.8.sentence-2) says that:

> [pointers to non-static class members](https://eel.is/c++draft/basic.compound#def:pointer_to_member),[32](https://eel.is/c++draft/basic.compound#footnote-32) which identify members of a given type within objects of a given class, [[dcl.mptr]](https://eel.is/c++draft/dcl.mptr)[.](https://eel.is/c++draft/basic.compound#1.8.sentence-1)
**Pointers to data members and pointers to member functions are collectively called [pointer-to-member](https://eel.is/c++draft/basic.compound#def:pointer-to-member) types[.](https://eel.is/c++draft/basic.compound#1.8.sentence-2)**

Note that the highlighted sentence 2 above does doesn't make distinction between static and non-static members and it just group all members(static as well as non-static) into the category of/called _pointer-to-member_ type. 

Now, we move onto [expr.unary.op](https://eel.is/c++draft/expr.unary.op#4) that states:

> A pointer to member is **only formed when an explicit** `&` is used and its operand is a [qualified-id](https://eel.is/c++draft/expr.prim.id.qual#nt:qualified-id) not enclosed in parentheses[.](https://eel.is/c++draft/expr.unary.op#4.sentence-1) .

This means that `int (*ptr_static)(int) = C::g;` in the below program is ill-formed because we haven't used `&`. But practically all compilers accept the program and we usually expect it to work because it `g` is a static member function.

```
struct C{
   static int g(int);
};

int (*ptr_static)(int) = C::g; //accepted by all compilers  but ill-formed as per current wording
 
```

Note that there is also [dcl.mptr](https://eel.is/c++draft/dcl.mptr#4) that says that a pointer to member shall not point to static member of a class:

> A pointer to member shall not point to a static member of a class ([[class.static]](https://eel.is/c++draft/class.static)), a member with reference type, or “cv void”[.](https://eel.is/c++draft/dcl.mptr#4.sentence-1)
>
> [Note [1](https://eel.is/c++draft/dcl.mptr#note-1): See also [[expr.unary]](https://eel.is/c++draft/expr.unary) and [[expr.mptr.oper]](https://eel.is/c++draft/expr.mptr.oper)[.](https://eel.is/c++draft/dcl.mptr#5.sentence-1)
The type “pointer to member” is distinct from the type “pointer”, that is, **a pointer to member is declared only by the pointer-to-member declarator syntax, and never by the pointer declarator syntax[.]**(https://eel.is/c++draft/dcl.mptr#5.sentence-2)
There is no “reference-to-member” type in C++[.](https://eel.is/c++draft/dcl.mptr#5.sentence-3)
— end note]

The important thing to note here is that this [dcl.mptr#4]  and the note above seems to imply that `ptr` is not a pointer to member. Note that this would justify the reason for the working of `C::g` above because we're not forming a pointer to member so  [expr.unary.op](https://eel.is/c++draft/expr.unary.op#4) isn't violated. But [basic.compound](https://eel.is/c++draft/basic.compound#1.8.sentence-2) said the pointer to data member and member functions are collectively called pointer-to-member types.

So for this first problem I suggest that [basic.compound#1.8.sentence-2] be changed to only include non-static members. Note that [basic.compound#1.3](https://eel.is/c++draft/basic.compound#1.3) already include static members.

----

This problem is even more evident/highlighted when considering  explicit object member function. For example, consider the following example for which we see [implementation divergence](https://godbolt.org/z/n1M5Wvofr):

```
struct C{
   int f(this int);
};

//both of these(#1 and #2) are accepted by msvc but rejected by clang and gcc
decltype(C::f) func;       //#1 well-formed as per current wording
int (*ptr)(int) = C::f;    //#2 ill-formed as per current wording


int main()
{ 
      func(3);     
 
}
```

In this program `#1` is well-formed as per current wording as explained [here](https://stackoverflow.com/questions/78739117/explicit-this-member-function-use-accepted-by-msvc-but-rejected-by-clang-and-gcc/78739143#78739143) but `#2` seems to be ill-formed because `C::f` can't be used to form a pointer to member as `&` is not used as per [expr.unary.op](https://eel.is/c++draft/expr.unary.op#4). Note that msvc accepts both `#1` and `#2`. 

Now as per my understanding, the explicit object member function is supposed to behave like a static member function. So I would expect `#2` to work as well even though the current wording makes it ill-formed. 

So maybe we also need to exempt explicit object member function from [basic.compound#1.8.sentence-2](https://eel.is/c++draft/basic.compound#1.8.sentence-2). 

Basically, ideally `C::f` should not behave like the old pointer to implicit object member function(and instead behave like a pointer `T`) and we don't have to use `&` before `C::f` to get a pointer from it.  It should also be able to decay implicitly just like normal functions do. So we might also need to do some modification in [conv.func].

<h4>Suggested resolution:</h4>

Change [basic.compound#1.8.sentence-2](https://eel.is/c++draft/basic.compound#1.8.sentence-2) to only include non-static data members and implicit object member function. Note [basic.compound#1.3](https://eel.is/c++draft/basic.compound#1.3) already include static members.

**Comment by AlanVoore** [comment]
So if I understand correctly, one of the problems that this issue deals with is that both `int (*ptr_static)(int) = C::g;` and `int (*ptr)(int) = C::f;` should be well-formed because both of these uses _pointer declarator syntax_ instead of _pointer to member declarator syntax_ and so the non-normative [note](https://eel.is/c++draft/dcl.mptr#note-1) cited in this issue is applicable to both of these declarations and we don't have to use `&` before either `C::f` as well as `C::g`. Meaning these should behave similarly/consistently. But [basic.compound#1.8](https://eel.is/c++draft/basic.compound#1.8.sentence-2) puts both static and non-static members to be under _pointer-to-member_ types. So we need to change [basic.compound#1.8](https://eel.is/c++draft/basic.compound#1.8.sentence-2) to only include non-static data member and implicit object member function. 

Is this correct interpretation of one of the problems raise in this issue?

**Comment by ranaanoop** [comment]
@AlanVoore > Is this correct interpretation of one of the problems raise in this issue?

Yes, exactly. 

In addition, this issue also points out that `decltype(C::f) func;` is already well-formed as per [current wording](https://eel.is/c++draft/expr.prim.id.general#4) which is the expected behavior because `decltype(C::f) func;` should behave same as `decltype(C::g) func;`. 

Baiscally, an explicit object member function is supposed to behave like a static member  function so they should be consistently handled.

**Comment by AlanVoore** [comment]
Here is one more demo showing the same problem as discussed in this issue:

```
struct C
{
  static int g(int);
  int f(this int);
};
template<typename T> void func(T&); 
int main()
{
    func(C::g); //accepted by all compiler & well-formed as per current wording
    func(C::f); //rejected by gcc and clang but well-formed as per current wording. MSVC accepts this 
}
```

Note `func(C::f)` is well-formed as per current wording but rejected by both clang and gcc and accepted by only msvc. Intuitionally, I would expected it to be valid as well. Note  `(&C::f)(44);` also compiles with gcc and clang and so it works like a static member function as no class type object was needed/passed here in `(&C::f)(44);`.

**Comment by jensmaurer** [comment]
Explicit object member functions cannot be named by a _qualified-id_ unless taking the address or when in a this-transformation context.  CWG2902 was amended accordingly. See #573.

Is there any remaining issue here, beyond that?
